lib/services/app.go

```diff
@@ -158,3 +168,222 @@ func UnmarshalAppServer(data []byte, opts ...MarshalOption) (types.AppServer, er
 	}
 	return nil, trace.BadParameter("unsupported app server resource version %q", h.Version)
 }
+
+// NewApplicationsFromKubeService creates application resources from kubernetes service. One service result
+// in multiple application resources if there are multiple ports exposed.
+func NewApplicationsFromKubeService(service v1.Service, clusterName string, pi ProtocolChecker) ([]types.Application, error) {
+	var apps types.Apps
+	protocol := ""
+	for k, v := range service.GetAnnotations() {
+		if k == types.DiscoveryProtocolLabel {
+			protocol = v
+		}
+	}
+
+	ports, err := getServicePorts(service)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	var appsMu sync.Mutex
+	g := errgroup.Group{}
+	g.SetLimit(5)
+	for _, port := range ports {
+		port := port
+		g.Go(func() error {
+			appURI := getAppURI(getServiceFQDN(service), port, protocol, pi)
+
+			// By default we are looking for http apps, we only create tcp apps if protocol annotation
+			// explicitly specifies it.
+			if strings.HasPrefix(appURI, protoTCP) && protocol != protoTCP {
+				return nil
+			}
+
+			rewriteConfig, err := getAppRewriteConfig(service.GetAnnotations())
+			if err != nil {
+				return trace.Wrap(err)
+			}
+			a, err := types.NewAppV3(types.Metadata{
+				Name:        getAppName(service.GetName(), service.GetNamespace(), clusterName, ""),
+				Description: fmt.Sprintf("Discovered application in Kubernetes cluster %q", clusterName),
+				Labels:      getAppLabels(service.GetLabels(), clusterName),
+			}, types.AppSpecV3{
+				URI: appURI,
+				// Temporary usage to have app's name with port name, in case there are more than one app
+				// created from this service.
+				PublicAddr: getAppName(service.GetName(), service.GetNamespace(), clusterName, port.Name),
+				Rewrite:    rewriteConfig,
+			})
+			if err != nil {
+				return trace.Wrap(err)
+			}
+			appsMu.Lock()
+			apps = append(apps, a)
+			appsMu.Unlock()
+			return nil
+		})
+	}
+	err = g.Wait()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// If we ended up with more than one app from the service, we'll have port name in the app name to distinguish them.
+	// We can only do this now because we don't know how many apps there will be until all ports are processed by errgroup.
+	for i := range apps {
```

Comment:
why don't you decide this before?
If ports>1....

Reply:
Because port can be skipped if it ends up being `tcp`. So, as comment says, we only know true amount of apps after all goroutines finish running.